### PR TITLE
#1692 - Ministry overaward part 2

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Ministry overaward part 2

- [x] In student profile, display a new tab called "Overawards"

image

- [x] Display a students overaward balance by CSLF, CSLP, BCSL

image

- [x] Display in a table a students overawards generated

image

- [x] Display in a table overawards deducted
image

- [x] Display a button called "Add manual record" triggering a modal to open with the type of award and dollar amount that the student paid back

image
image
image

- [x] The value entered into the "Add manual record" modal, must be inverted when inserting into the table.
image
image

- [x] Validation for value entered into the "Add manual record" - Between 0 - 99999
image

- [x] All displays to be created as common components (will be eventually reused in Student Portal)

### CODE REVIEW

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          type="number"
+          prefix="$"
+          :rules="[
+            (v) => checkNullOrEmptyRule(v, 'Overaward deduction amount'),
+            (v) =>
+              numberRangeRule(
+                v,
+                -MONEY_VALUE_FOR_UNKNOWN_MAX_VALUE,
+                MONEY_VALUE_FOR_UNKNOWN_MAX_VALUE,
+                'Overaward deduction amount',
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:33]: The `formatCurrency` can be added here to have the numbers formatted as "money".
If possible, remove the "(CAD)" from the `formatCurrency` method.

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+              ),
+          ]"
+        />
+      </template>
+      <template #footer>
+        <check-permission-role :role="allowedRole">
+          <template #="{ notAllowed }">
+            <footer-buttons
+              primaryLabel="Add Record now"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:42]: The case should probably be "Add Record Now" or "Add record now".

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                -MONEY_VALUE_FOR_UNKNOWN_MAX_VALUE,
+                MONEY_VALUE_FOR_UNKNOWN_MAX_VALUE,
+                'Overaward deduction amount',
+              ),
+          ]"
+        />
+      </template>
+      <template #footer>
+        <check-permission-role :role="allowedRole">
+          <template #="{ notAllowed }">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:40]: I am aware that it works and we used it before in other places but I am now getting the below error. Has anyone else faced the same? I am using `ESLint v2.4.0`, `Vetur v0.37.3`, `Prettier v9.10.4`
image

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    },
+  },
+  setup() {
+    const { numberRangeRule, checkNullOrEmptyRule } = useRules();
+    const { showDialog, showModal, resolvePromise } = useModalDialog<
+      OverawardManualRecordAPIInDTO | boolean
+    >();
+    const addManualOverawardForm = ref({} as VForm);
+    const formModel = reactive({} as OverawardManualRecordAPIInDTO);
+    const awardTypeItems = [...FULL_TIME_AWARDS, ...PART_TIME_AWARDS]
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:83]: IMO, a simple `select` component with the 3 hardcoded values would be enough.
If we are planning to have `FULL_TIME_AWARDS` and `PART_TIME_AWARDS` as a source for loans/grants I would suggest the below.

-  `FULL_TIME_AWARDS` and `PART_TIME_AWARDS` can be combined in one single list and Full-Time/Part-Time can be a property.
- More properties/metadata can be added to make the filter easier, for instance, `allowOverawardEntry` can be added to `CSLF, CSLP, BCSL` making the filtering less hard-coded.

Comment (andrewsignori-aot) [reply sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:83]: I saw now from where it came in the refactor on the PR 😉

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+              <template #body="slotProps">
+                <span>
+                  {{ dateOnlyLongString(slotProps.data.dateAdded) }}
+                </span>
+              </template>
+            </Column>
+            <Column field="applicationNumber" header="Application #"
+              ><template #body="slotProps">
+                <span>
+                  {{ slotProps.data.applicationNumber ?? "-" }}
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:27]: I believe that we have the `emptyStringFiller` for this.

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                </span>
+              </template></Column
+            >
+            <Column field="overawardOrigin" header="Origin"></Column>
+            <Column field="assessmentTriggerType" header="Type"></Column>
+            <Column field="awardValueCode" header="Award"></Column>
+            <Column field="overawardValue" header="Overaward amount">
+              <template #body="slotProps">
+                <span>
+                  {{ formatCurrency(slotProps.data.overawardValue, "") }}
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:37]: We can remove the `currencyCode = "(CAD)"` from the `formatCurrency` as per the last conversations with @JasonCTang and @hellolynn-tbtb, and then the last empty string parameter will not be required here.

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+                <span>
+                  {{ slotProps.data.applicationNumber ?? "-" }}
+                </span>
+              </template></Column
+            >
+            <Column field="overawardOrigin" header="Origin"></Column>
+            <Column v-if="showAddedBy" field="addedByUser" header="Added By"
+              ><template #body="slotProps">
+                <span>
+                  {{ slotProps.data.addedByUser ?? "-" }}
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:95]: Same form https://github.com/bcgov/SIMS/pull/1748/files#r1116087956

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    const addManualOverawardDeduction = ref(
+      {} as ModalDialog<OverawardManualRecordAPIInDTO | boolean>,
+    );
+    const overawards = computed(() =>
+      overawardDetails.value.filter(
+        (overaward) =>
+          ![
+            DisbursementOverawardOriginType.AwardDeducted,
+            DisbursementOverawardOriginType.ManualRecord,
+          ].includes(overaward.overawardOrigin),
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:171]: Can we have the array added to a const and shared between `overawards` and `overawardDeductions` filters?

File: sources/packages/web/src/components/common/OverawardDetails.vue (hunk 6 above)
Comment (ann-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:37]: why second parameter is `""`?

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+      const manualOveraward =
+        await addManualOverawardDeduction.value.showModal();
+      if (manualOveraward) {
+        try {
+          await OverawardService.shared.addManualOverawardDeduction(
+            props.studentId,
+            manualOveraward as OverawardManualRecordAPIInDTO,
+          );
+          snackBar.success("Overaward deduction added successfully.");
+          context.emit("updateOverawardDetails");
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:194]: I believe that the event must be named after what it represents, for instance, `overawardBalanceUpdated` or `manualOverawardAdded`. `updateOverawardDetails` is meaningful for the consumer of the event only.

File: sources/packages/web/src/components/common/StudentOverawardDetails.vue [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+      />
+      <v-row>
+        <v-col>
+          <title-value
+            :propertyValue="
+              formatCurrency(overawardBalance.overawardBalanceValues?.CSLF, '')
+            "
+          >
+            <template #title>
+              CSLF
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/StudentOverawardDetails.vue:16]: On the Vue side, the loan/grants acronyms can be defined in a const to also be used by `FULL_TIME_AWARDS/PART_TIME_AWARDS`. Looking more into the code, I have a stronger feeling that the below comment makes sense (the part of the bullet).
https://github.com/bcgov/SIMS/pull/1748/files#r1116077791

File: sources/packages/web/src/composables/useFormatters.ts [hunk 11]
```diff
@@ -210,19 +210,26 @@ export function useFormatters() {
...
    * @param currencyCode currency code.
+   * @param defaultValueAmount default value to display
+   * when value is not provided or zero.
    * @returns formatted currency value.
    */
   const formatCurrency = (
     moneyAmount: number | string,
     currencyCode = "(CAD)",
-  ): string | null => {
+    defaultValueAmount = "$0.00",
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/composables/useFormatters.ts:223]: I would prefer to have 0(zero) assigned to the `moneyAmount` and let the format be done by the method.
The consumer of the method can also do it using the `??` for the cases where it makes sense.

File: sources/packages/web/src/components/common/StudentOverawardDetails.vue [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    <v-container>
+      <body-header
+        title="Overaward balances"
+        subTitle="A balance of overawards broken down by award type"
+      />
+      <v-row>
+        <v-col>
+          <title-value
+            :propertyValue="
+              formatCurrency(overawardBalance.overawardBalanceValues?.CSLF, '')
```
Comment (ann-aot) [inline sources/packages/web/src/components/common/StudentOverawardDetails.vue:12]: will `overawardBalanceValues` be undefined in any case? `formatCurrency` first parameter is `number | string`, its not expecting `undefined`.
image

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/web/src/components/common/OverawardDetails.vue (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/OverawardDetails.vue:37]: @ann-aot please see https://github.com/bcgov/SIMS/pull/1748#discussion_r1116090487

File: sources/packages/web/src/components/generic/TitleValue.vue [hunk 13]
```diff
@@ -14,13 +16,13 @@ Used when we need to display title and the value inside a container
 import { defineComponent } from "vue";
 export default defineComponent({
   props: {
-    propertyTitle: {
+    propertyValue: {
       type: String,
       required: true,
     },
-    propertyValue: {
+    propertyTitle: {
```
Comment (ann-aot) [inline sources/packages/web/src/components/generic/TitleValue.vue:23]: why this change?

File: sources/packages/web/src/components/common/StudentOverawardDetails.vue (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/StudentOverawardDetails.vue:12]: `formatCurrency` `moneyAmount` parameter can be changed to accept `undefined`.

File: sources/packages/web/src/composables/useFormatters.ts [hunk 14]
```diff
@@ -210,19 +210,26 @@ export function useFormatters() {
...
+  ): string => {
     if (moneyAmount === null || moneyAmount === undefined) {
-      return null;
+      return defaultValueAmount;
     }
-    return `$${parseFloat(moneyAmount.toString()).toFixed(2)} ${currencyCode}`;
+    const currencyValue = parseFloat(moneyAmount.toString());
+    const negativeValueDisplay = currencyValue < 0 ? "-" : "";
```
Comment (ann-aot) [inline sources/packages/web/src/composables/useFormatters.ts:229]: mm... i didn't get the need of this logic

File: sources/packages/web/src/views/aest/student/Overawards.vue [hunk 15]
```diff
@@ -0,0 +1,23 @@
+<template>
+  <p class="header-sub-title mt-9">Overawards</p>
```
Comment (ann-aot) [inline sources/packages/web/src/views/aest/student/Overawards.vue:2]: it's not matching the other tabs 
image
image

File: sources/packages/web/src/composables/useFormatters.ts (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/composables/useFormatters.ts:229]: It is to display "-$99.99" instead of "$-99.99". I hope that it is what Ministry expects also 😄.
@hellolynn-tbtb FYI

File: sources/packages/web/src/constants/award-constants.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
+export interface AwardDetail {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/constants/award-constants.ts:1]: Just a suggestion.
```ts
export interface AwardDetail {
  awardType: FullTimeAwardsTypes | PartTimeAwardsTypes;
  description: string;
}

export enum FullTimeAwardsTypes {
  CSLF,
  CSGP,
  CSGD,
  CSGF,
  CSGT,
  BCSL,
  BCAG,
  BGPD,
  SBSD,
}

export enum PartTimeAwardsTypes {
  CSLP,
  CSGP,
  CSPT,
  CSGD,
  BCAG,
  SBSD,
}
```

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue (hunk 3 above)
Comment (ann-aot) [reply sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:40]: I am not getting it
image
 
 I am using ESLint v2.4.0, Vetur v0.37.3, Prettier v9.10.4ttps://user-images.githubusercontent.com/77353155/221025392-9ad10d3b-c0d0-498c-af3a-91dd00f6d150.png)

File: sources/packages/web/src/composables/useFormatters.ts (hunk 14 above)
Comment (ann-aot) [reply sources/packages/web/src/composables/useFormatters.ts:229]: oh..okay

**Review by ann-aot — COMMENTED** [review]
Good work @dheepak-aot 👍 added some comments.

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:40]: I am also getting the same error everywhere #="{ notAllowed }" is used. 
But it doesn't show up in the npm run serve console.

image

Comment (andrewsignori-aot) [reply sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:40]: @ann-aot did you check the extensions versions? Are they the same ones that you are using?

File: sources/packages/web/src/components/common/StudentOverawardDetails.vue (hunk 12 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/common/StudentOverawardDetails.vue:12]: If it is expecting undefined, then it will be like an optional parameter right? it's consumers may use it like formatCurrency().

IMO like it as is will be better.

File: sources/packages/web/src/components/generic/TitleValue.vue (hunk 13 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/generic/TitleValue.vue:23]: In props propertyTitle was moved after propertyValue because propertyTitle is not a required property anymore because of the slot provided to create title.

File: sources/packages/web/src/views/aest/student/Overawards.vue (hunk 15 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/aest/student/Overawards.vue:2]: The notes title is not correct as per figma.

This is how it is 

image

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue (hunk 3 above)
Comment (ann-aot) [reply sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:40]: ya @andrewsignori-aot Iam using the same ESLint v2.4.0, Vetur v0.37.3, Prettier v9.10.4

**Review by andrepestana-aot — APPROVED** [review]
Great job.

File: sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/aest/students/modals/AddManualOverawardDeduction.vue:41]: I missed the details on your comment @ann-aot. 
Either way, I would prefer to stop using the syntax.

File: sources/packages/web/src/components/common/OverawardDetails.vue [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+              <template #body="slotProps">
+                <span>
+                  {{ dateOnlyLongString(slotProps.data.dateAdded) }}
+                </span>
+              </template>
+            </Column>
+            <Column field="applicationNumber" header="Application #">
+              <template #body="slotProps">
+                <span>
+                  {{ slotProps.data.applicationNumber ?? emptyStringFiller() }}
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/OverawardDetails.vue:87]: The intention is to use it as below.
```ts
{{ emptyStringFiller(slotProps.data.applicationNumber) }}
```

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes and the beyond-the-pr refactors 😉 
Looks good, just have one last minor comment.

File: sources/packages/web/src/components/common/OverawardDetails.vue (hunk 17 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/common/OverawardDetails.vue]: Got it. done.

**Review by andrewsignori-aot — APPROVED** [review]

File: sources/packages/web/src/views/aest/student/Overawards.vue (hunk 15 above)
Comment (ann-aot) [reply sources/packages/web/src/views/aest/student/Overawards.vue:2]: mm... I think figma may have updated after the latest refactor. but we follow the format, right? in the notes there is a `div` container

File: sources/packages/web/src/components/generic/TitleValue.vue (hunk 13 above)
Comment (ann-aot) [reply sources/packages/web/src/components/generic/TitleValue.vue:23]: an order doesn't matter here right?

**Review by ann-aot — APPROVED** [review]
Good work @dheepak-aot 👍
